src/launcher/plugin.ts

```diff
@@ -0,0 +1,226 @@
+
+
+  app.commands.addCommand('jupyterlab-launcher:show', {
+    label: 'Show Launcher',
+    execute: () => {
+      if (!launcherWidget.isAttached) {
+        app.shell.addToLeftArea(launcherWidget);
+      } else {
+        app.shell.activateMain(launcherWidget.id);
+      }
+    }
+  });
+
+  palette.addItem({
+    command: 'jupyterlab-launcher:show',
+    category: 'Help'
+  });
+
+  app.shell.addToLeftArea(launcherWidget);
+}
+
+
+/**
+ * Simple encapsulation of name and callback of launcher entries.
+ */
+class LauncherItem {
+  constructor(public name: string, public clickCallback: () => void) {
+    }
+}
+
+
+/**
+ * LauncherModel keeps track of the path to working directory and has a list of
+ * LauncherItems.
+ */
+class LauncherModel  extends VDomModel {
+  items : LauncherItem[] = [];
+  path : string = 'home';
+  /**
+   * Convenience method to add a launcher with a given name and callback.
+   *
+   * This keeps plugins who wish to register themselves in the Launcher from
+   * having to import both LauncherItems and LauncherItem.
+   */
+  add(name: string, clickCallback: () => void) : IDisposable {
+    return this.addItem(new LauncherItem(name, clickCallback));
+  }
+
+  /**
+   * Add a new launcher and trigger re-render event for parent widget.
+   *
+   * returns a Disposable which can be called to remove this new item from
+   * the Launcher and trigger another re-render event.
+   */
+  addItem(item: LauncherItem) : IDisposable {
+    this.items.push(item);
+    this.stateChanged.emit(void 0);
+
+    return new DisposableDelegate(() => {
+      // Remove the item form the list of items.
+      var index = this.items.indexOf(item, 0);
+      if (index > -1) {
+          this.items.splice(index, 1);
+          this.stateChanged.emit(void 0);
+      }
+    });
+  }
+  
+  setDir(path: string) : void {
+    this.path = path;
+    this.stateChanged.emit(void 0);
+  }
+}
+
+
+/**
+ * A virtual-DOM-based widget for the Launcher
+ */
+class LauncherWidget extends VDomWidget<LauncherModel> {
+
+  protected render(): VNode | VNode[] {
+    let children : VNode[] = [];
+
+    for (let item of this.model.items) {
+      let imgName = item.name.replace(' ', '');
+      let img = h.span({className: `jp-Image${imgName} jp-Launcher-image`});
+      let text = h.span({className: 'jp-Launcher-text' }, item.name);
+
+      let column = h.div({
+        className: 'jp-Launcher-column',
+        'onclick': item.clickCallback
+      }, [img, text])
+      children.push(column);
+    }
+
+    let folderImage = h.span({ className: 'jp-Launcher-folder' });
+    let p = this.model.path;
+    let pathName = p.length ? 'home > ' + p.replace('/', ' > ') : 'home';
+    let path = h.span({ className: 'jp-Launcher-path' }, pathName );
+
+    let cwd = h.div({ className: 'jp-Launcher-cwd' }, [folderImage, path]);
+    let body = h.div({ className: "jp-Launcher-body" }, children);
+
+    return h.div({ className: 'jp-Launcher-dialog'}, [ cwd, body ]);
```

Comment:
This return `div` seems like an extra level of nesting. I think it would be better to just `return [cwd, body];`